src/Portal.js

```diff
@@ -13,20 +13,81 @@ let Portal = React.createClass({
     container: CustomPropTypes.mountable
   },
 
-  // we use the mixin for now, to avoid duplicating a bunch of code.
-  // when the deprecation is removed we need to move the logic here from OverlayMixin
-  mixins: [ OverlayMixin ],
+  componentDidMount() {
+    this._renderOverlay();
+  },
+
+  componentDidUpdate() {
+    this._renderOverlay();
+  },
+
+  componentWillUnmount() {
+    this._unrenderOverlay();
+    this._mountOverlayTarget();
+  },
+
+  _mountOverlayTarget() {
+    if (!this._overlayTarget) {
+      this._overlayTarget = document.createElement('div');
+      this.getContainerDOMNode()
+        .appendChild(this._overlayTarget);
+    }
+  },
+
+  _unmountOverlayTarget() {
+    if (this._overlayTarget) {
+      this.getContainerDOMNode()
+        .removeChild(this._overlayTarget);
+      this._overlayTarget = null;
+    }
+  },
+
+  _renderOverlay() {
 
-  renderOverlay() {
-    if (!this.props.children) {
-      return null;
+    let overlay = !this.props.children
+      ? null
+      : React.Children.only(this.props.children);
+
+    // Save reference for future access.
+    if (overlay !== null) {
+      this._mountOverlayTarget();
+      this._overlayInstance = React.render(overlay, this._overlayTarget);
+    } else {
+      // Unrender if the component is null for transitions to null
+      this._unrenderOverlay();
+      this._unmountOverlayTarget();
     }
+  },
 
-    return React.Children.only(this.props.children);
+  _unrenderOverlay() {
+    if (this._overlayTarget) {
+      React.unmountComponentAtNode(this._overlayTarget);
+      this._overlayInstance = null;
+    }
   },
 
   render() {
     return null;
+  },
+
+  getOverlayDOMNode() {
+    if (!this.isMounted()) {
```

Comment:
I only looked at the partial diff and saw `getOverlayDOMNode() {`, I always forget about shorthand function properties...